### PR TITLE
Beam Rifle Zooming Rework + Click catcher memes + bunch of other random stuff that players won't use

### PR DESCRIPTION
**Pull request by silicons**
- Click catcher now supports mousemove/mousedrag. Basically, mouse location can now be captured even if the user isn't mousing over a visible location
- Added procs to allow for projectiles to be fired with just an angle for pixel projectiles, instead of requiring a target turf and pixel x/y
- Added procs to get angle of user's mouse from their viewpoint (Time to rework gang machine guns again!)
- Beam rifles now have different zoom modes
- Free directional zooms out and tracks the angle of your mouse from the center of the screen. However, you can't target with very good accuracy on this (Shots can't be properly aimed on non dense objects/lying down mobs.)
- Locked directional zooms like free directional but doesn't automatically turn if your aim changes.
- Center view, just increases your view in all directions (2x weaker)
- No zoom mode, in which you just retain your normal view.
- You can select beam rifle zooming rates to be instant or stepped. 
- Stepped zooming rates zoom out 5 tiles per second. This will likely help with people not being able to use it without lagging because their computers aren't as beefy!
- Beam rifles no longer require zoom to be fired
- Beam rifle aiming beams now instantly update instead of on process
- Beam rifle aiming beams are now one object instead of 150. This'll help with the lag caused by it during gameplay that I've observed.
- Angular penalty reduced by 0.1 for a nice even number.
- Instances of client.view = <view> have been replaced with client.change_view() as that'll properly update the click catcher
- Hopefully shooting yourself in the face when you hit a blob tile or whatnot is fixed with the new and improved code..
Props to @RemieRichards and @MrPerson!

### CODE REVIEW

**Comment by Repukan** [comment]
Freeze

**Comment by Cyberboss** [comment]
This PR feels mighty close to the freezing point

**Comment by Repukan** [comment]
When in doubt close

**Comment by optimumtact** [comment]
it's a violation of good taste in atomising pr's but I don't think it adds any features that are antag only

**Comment by kevinz000** [comment]
@optimumtact i was told to not do things until i need them hence the clickcatcher memes and other procs.

**Comment by kevinz000** [comment]
merge token from #28592 @korphaeron

**Comment by KorPhaeron** [comment]
Sure

File: code/_onclick/autoclick.dm [hunk 1]
```diff
@@ -6,6 +6,9 @@
 	var/mouseObject = null
 	var/mouseControlObject = null
 
+/mob
+	var/list/mousemove_intercept_objects
```
Comment (lzimann) [inline code/_onclick/autoclick.dm]: Put this var in `mob_defines.dm`

File: code/modules/projectiles/guns/beam_rifle.dm [hunk 2]
```diff
@@ -85,149 +180,173 @@
...
 	. = ..()
 	START_PROCESSING(SSfastprocess, src)
+	zoom_speed_action = new(src)
+	zoom_lock_action = new(src)
 
 /obj/item/weapon/gun/energy/beam_rifle/Destroy()
 	STOP_PROCESSING(SSfastprocess, src)
+	set_user(null)
+	clear_tracer()
 	..()
```
Comment (lzimann) [inline code/modules/projectiles/guns/beam_rifle.dm]: `return ..()`

File: code/modules/projectiles/guns/beam_rifle.dm [hunk 3]
```diff
@@ -85,149 +180,173 @@
...
-	if(scoped)
+	if(aiming)
 		slowdown = scoped_slow
 	else
 		slowdown = initial(slowdown)
 
 /obj/item/weapon/gun/energy/beam_rifle/Initialize()
 	. = ..()
 	START_PROCESSING(SSfastprocess, src)
+	zoom_speed_action = new(src)
```
Comment (lzimann) [inline code/modules/projectiles/guns/beam_rifle.dm]: These actions are never being granted

File: code/modules/projectiles/guns/beam_rifle.dm [hunk 4]
```diff
@@ -85,149 +180,173 @@
...
-		return
-	var/turf/T = get_turf(current_user.client.mouseObject)
-	if(!istype(T))
+/obj/item/weapon/gun/energy/beam_rifle/proc/aiming_beam(force_update = FALSE)
+	var/diff = abs(aiming_lastangle - lastangle)
+	check_user()
+	if(diff < AIMING_BEAM_ANGLE_CHANGE_THRESHOLD && !force_update)
 		return
+	aiming_lastangle = lastangle
+	CHECK_TICK
```
Comment (lzimann) [inline code/modules/projectiles/guns/beam_rifle.dm]: Why is this `CHECK_TICK` here?

Comment (kevinz000) [inline code/modules/projectiles/guns/beam_rifle.dm]: leftovers from before when i tried making 1500 objects per second and felt it was laggy but it's not needed now i'll remove it

File: code/modules/projectiles/guns/beam_rifle.dm [hunk 5]
```diff
@@ -85,149 +180,173 @@
...
+	var/turf/curloc = get_turf(src)
+	var/turf/targloc = get_turf(current_user.client.mouseObject)
+	if(!istype(targloc))
+		if(!istype(curloc))
+			return
+		targloc = get_turf_in_angle(lastangle, curloc, 10)
+	P.preparePixelProjectile(targloc, targloc, current_user, current_user.client.mouseParams, 0)
+	P.fire(lastangle)
+
+/obj/item/weapon/gun/energy/beam_rifle/proc/clear_tracer()
```
Comment (lzimann) [inline code/modules/projectiles/guns/beam_rifle.dm]: Feels really wasteful having an extra proc just to call `QDEL_NULL(current_tracer)`. Either just use qdel null directly or make it a define.

File: code/modules/projectiles/guns/beam_rifle.dm (hunk 3 above)
Comment (kevinz000) [inline code/modules/projectiles/guns/beam_rifle.dm]: they are i've tested they work

File: code/_onclick/click.dm [hunk 6]
```diff
@@ -432,28 +432,33 @@
 
 /obj/screen/click_catcher
 	icon = 'icons/mob/screen_gen.dmi'
-	icon_state = "click_catcher"
+	icon_state = "flash"
```
Comment (optimumtact) [inline code/_onclick/click.dm]: why did you change the icon_state ?

**Comment by optimumtact** [comment]
zooming and tracing could both be helper datums probably, anyway this looks fine but you should fix the conflicts

File: code/_onclick/click.dm (hunk 6 above)
Comment (kevinz000) [inline code/_onclick/click.dm]: woops

File: code/modules/projectiles/guns/beam_rifle.dm [hunk 7]
```diff
@@ -37,13 +40,14 @@
...
+	var/aiming_time_increase_angle_multiplier = 0.3
 
 	var/lastangle = 0
+	var/aiming_lastangle = 0
 	var/mob/current_user = null
-	var/list/obj/effect/temp_visual/current_tracers = list()
+	var/obj/effect/projectile_beam/current_tracer
 
-	var/structure_piercing = 2				//This doesn't always work!
+	var/structure_piercing = 4				//Amount * 2. For some reason structures aren't respecting this unless you have it doubled. Probably with the objects in question's Bump() code instead of this but I'll deal with this later.
```
Comment (optimumtact) [inline code/modules/projectiles/guns/beam_rifle.dm]: ermm

Comment (optimumtact) [inline code/modules/projectiles/guns/beam_rifle.dm]: it might be bumping against the turf, or itself, or perhaps it bumps on enter and then on exit

File: code/modules/projectiles/guns/beam_rifle.dm [hunk 8]
```diff
@@ -355,16 +489,21 @@
...
+					var/turf/closed/wall/W = target
+					W.dismantle_wall(TRUE, TRUE)
+				else
+					target.ex_act(2)
 			return TRUE
 	if(ismovableatom(target))
 		var/atom/movable/AM = target
 		if(AM.density && !AM.CanPass(src, get_turf(target)) && !ismob(AM))
-			if(structure_pierce++ < structure_pierce_amount)
+			if(structure_pierce < structure_pierce_amount)
```
Comment (optimumtact) [inline code/modules/projectiles/guns/beam_rifle.dm]: did it work

**Comment by kevinz000** [comment]
@optimumtact pierce is fixed
I'm still debugging the pickup bugs people have reported.

**Comment by kevinz000** [comment]
fixed, I had an infinite loop in it..

**Comment by optimumtact** [comment]
kev you're conflicting

**Comment by kevinz000** [comment]
fixed

**Comment by optimumtact** [comment]
are you ready for review again?

**Comment by kevinz000** [comment]
IM FIIIXXIIIINGGG

**Comment by kevinz000** [comment]
Yes I am
